### PR TITLE
Add a hamiltonian rule for the jax operators

### PR DESCRIPTION
**Pull request by inailuig**
This is what is left of https://github.com/netket/netket/pull/968 rebased on master.
closes  https://github.com/netket/netket/pull/968.

Makes the hamiltonian rule work on gpu as no numba4jax is required anymore.
@PhilipVinc please tell me we don't need to use dispatch to select it automatically based on the type of operator.

### CODE REVIEW

File: netket/sampler/rules/hamiltonian.py [hunk 1]
```diff
@@ -114,3 +116,70 @@ def _choose(vp, sections, rand_vec, out, w):
...
+        xp, mels = self.operator.get_conn_padded(x)
+
+        n_conn = self.operator.n_conn(x)
+
+        # def _n_conn(mels):
+        #     nonzeros = jnp.abs(mels) > 0
+        #     return nonzeros.sum(axis=-1)
+        # n_conn = _n_conn(mels)
+
+        rand_i = jax.random.randint(key, shape=(x.shape[0],), minval=0, maxval=n_conn)
```
Comment (inailuig) [inline netket/sampler/rules/hamiltonian.py:146]: here it would be useful to have  some kind of `JaxOperator.get_conn_padded_and_n_conn` as in some of the operators to come we compute the n_conn inside of get_conn_padded anyway.

(here n_conn tells us how many nonzeros there are at the beginning of the mels of each sample, the rest is assumed to be zero padding, which would not have been present in the old get_conn_flattened)

File: netket/sampler/rules/hamiltonian.py [hunk 2]
```diff
@@ -114,3 +116,70 @@ def _choose(vp, sections, rand_vec, out, w):
...
+        #
+        # if we were to assume that the nonzero mels are gathered at the beginning of each
+        # row we could avoid this, i.e. just do
+        # x_proposed = xp[jnp.arange(xp.shape[0]), rand_i]
+        #
+        # instead of actually shifting we build a mask for the selected element in each row
+        # and sum over the zeros at the end
+
+        # TODO let the operator supply the nonzeros or nonzero_i_plus1 directly
+        nonzeros = jnp.abs(mels) > 0
```
Comment (inailuig) [inline netket/sampler/rules/hamiltonian.py:158]: alternatively it would be even more useful if the operator already gave us a mask telling us where the nonzeros are, not just how many there are, then we could avoid checking here

Comment (PhilipVinc) [reply netket/sampler/rules/hamiltonian.py:158]: can you propose an API about this that would make sense? Would this be useful somewhere else as well?

File: netket/sampler/rules/hamiltonian.py [hunk 3]
```diff
@@ -114,3 +116,70 @@ def _choose(vp, sections, rand_vec, out, w):
...
         low_range = s
+
+
+@struct.dataclass
+class HamiltonianRuleJax(HamiltonianRule):
+    """
+    Rule proposing moves according to the terms in an operator.
+
+    In this case, the transition matrix is taken to be:
+
```
Comment (PhilipVinc) [inline netket/sampler/rules/hamiltonian.py:127]: need doctoring explaining this only works with jax friendly operators

**Comment by PhilipVinc** [comment]
Not dispatch, but I wouldn't mind if `nk.sampler.rules.HamiltonianRule` was a thin function automatically selecting between `HamiltonianRuleNumba` and `HamiltonianRuleJax` maybe by checking `isinstance(operator, DiscreteJaxOperator)`.

File: netket/sampler/rules/hamiltonian.py (hunk 3 above)
Comment (inailuig) [reply netket/sampler/rules/hamiltonian.py:137]: its 5 lines below

File: netket/sampler/rules/hamiltonian.py (hunk 2 above)
Comment (inailuig) [reply netket/sampler/rules/hamiltonian.py:168]: the simplest thing I could think of is to add two flags to get_conn_padded, return_n_conn and return_mask, then the operator can decide what he wants to return, with None being not available.
Then depending of the application we request what we want work with what we get.

could be useful in the future if we have operators which have very non-uniform number of connected elements, or really dont know a priory how many there are, e.g. in chemistry. We could then do the computations in a loop and stop when we reached the number of connected elements actually found.
 Also would be useful for checking we did not find more connected elements than _max_conn_size (applies to e.g. chemistry)

I think we can save it for a later PR though, once we have those operators in jax it will be more clear.

File: netket/sampler/rules/__init__.py [hunk 4]
```diff
@@ -17,7 +17,7 @@
 from .fixed import FixedRule
 from .local import LocalRule
 from .exchange import ExchangeRule
-from .hamiltonian import HamiltonianRule
+from .hamiltonian import HamiltonianRule, HamiltonianRuleJax
```
Comment (inailuig) [inline netket/sampler/rules/__init__.py:20]:
```suggestion
from .hamiltonian import HamiltonianRule
```

File: netket/sampler/rules/hamiltonian.py [hunk 5]
```diff
@@ -114,3 +126,86 @@ def _choose(vp, sections, rand_vec, out, w):
...
+def HamiltonianRule(operator):
+    """
+    Rule proposing moves according to the terms in an operator.
+
+    In this case, the transition matrix is taken to be:
+
+    .. math::
+
+       T( \\mathbf{s} \\rightarrow \\mathbf{s}^\\prime) = \\frac{1}{\\mathcal{N}(\\mathbf{s})}\\theta(|H_{\\mathbf{s},\\mathbf{s}^\\prime}|),
+    """
```
Comment (PhilipVinc) [inline netket/sampler/rules/hamiltonian.py:207]:
```suggestion

    This is a thin wrapper on top of the constructors of :class:`netket.sampler.rules.HamiltonianRuleJax` and
    :class:`netket.sampler.rules.HamiltonianRuleNumba`, which dispatches on one of the two implementations
    depending on whether the operator specified is jax-compatible (:class:`netket.operator.DiscreteJaxOperator`) 
    or not.
    """
```

**Review by PhilipVinc — APPROVED** [review]

**Comment by PhilipVinc** [comment]
Looks good, thanks!
Squash and merge once you're ready.

**Comment by PhilipVinc** [comment]
@inailuig ah please add a line to the changelong as well.

**Comment by PhilipVinc** [comment]
@inailuig can I merge?
